orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/servergroup/ServerGroupCacheForceRefreshTask.groovy

```diff
@@ -26,43 +29,164 @@ import groovy.util.logging.Slf4j
-        } catch (e) {
-          if (!outputs.containsKey("force.cache.refresh.errors")) {
-            outputs["force.cache.refresh.errors"] = []
+
+        return !stageData.refreshedServerGroups.contains(model) ? model : null
+      }
+    }.flatten()
+
+    if (!refreshableServerGroups) {
+      return Optional.empty()
+    }
+
+    def status = RUNNING
+    refreshableServerGroups.each { Map<String, String> model ->
+      try {
+        def response = oort.forceCacheUpdate(cloudProvider, REFRESH_TYPE, model)
+        if (response.status == HttpURLConnection.HTTP_OK) {
+          // cache update was applied immediately, no need to poll for completion
+          status = SUCCEEDED
+        }
+
+        stageData.refreshedServerGroups << model
+      } catch (e) {
+        stageData.errors << e.message
+      }
+    }
+    return Optional.of(new DefaultTaskResult(status, objectMapper.convertValue(stageData, Map)))
+  }
+
+  /**
+   * Ensure that:
+   * - We see a pending force cache update for every deployed server group
+   * - The pending force cache update is recent (newer than the start of this particular stage)
+   * - The pending force cache update has been processed (ie. it's survived one full pass of the caching agent)
+   *
+   * It's possible waiting until processing is overkill but we do this to avoid the possibility of a race condition
+   * between a forceCache refresh and an ongoing caching agent cycle.
+   */
+  private boolean processPendingForceCacheUpdates(String account,
+                                                  String cloudProvider,
+                                                  StageData stageData,
+                                                  Long startTime) {
+    def pendingForceCacheUpdates = oort.pendingForceCacheUpdates(cloudProvider, REFRESH_TYPE)
+
+    boolean finishedProcessing = true
+    stageData.deployServerGroups.each { String region, Set<String> serverGroups ->
+      finishedProcessing = finishedProcessing && serverGroups.every { String serverGroup ->
+        def model = [
+          serverGroup: serverGroup, region: region, account: account
+        ]
+
+        def forceCacheUpdate = pendingForceCacheUpdates.find {
+          (it.details as Map<String, String>).intersect(model) == model
+        }
+
+        if (!forceCacheUpdate) {
+          // there is no pending cache update, force it again in the event that it was missed
+          stageData.removeRefreshedServerGroup(model.serverGroup, model.region, model.account)
+          log.warn("Unable to find pending cache refresh request (model: ${model})")
+
+          return false
+        }
+
+        if (forceCacheUpdate) {
+          if (!forceCacheUpdate.processedTime || !forceCacheUpdate.cacheTime) {
+            // there is a pending cache update that is still awaiting processing
+            log.warn("Awaiting processing on pending cache refresh request (model: ${model})")
```

Comment:
Possibly ... I'd like to leave it in initially in case something doesn't go quite right, the added context would be beneficial.